Added Logic App Extension to stable api version

Shubhrmcf-07 opened the pull request:
# Logic Apps Extension on Microsoft.App RP

This PR aims to introduce logic app capabilities on top of container apps as a new offering. We already offer our services on Microsoft.Web RP and introducing a new SKU here. 

This spec is already a part of the preview API versions and is released to the customers as Logic App (Hybrid) offering. 

Key APIs introduced : 

- Logic App List Connections : Gets the connections for a particular logic app
- Logic App Get Workflows : Lists the workflows present on the app.
- Logic App Get Workflow : Get the workflow properties for a specific workflow on the app

Shubhrmcf-07 committed:
Added Logic App Extension to stable api version

Shubhrmcf-07 committed:
Added Logic App Extension to stable api version

Shubhrmcf-07 committed:
Testing lintdiff build

Shubhrmcf-07 committed:
Testing lintdiff build

Shubhrmcf-07 committed:
Lintdiff build fix

Shubhrmcf-07 committed:
Testing lintdiff build

razvanbadea-msft commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,742 @@
+        "tags": [
+          "LogicApps"
+        ],
+        "summary": "Proxies a the API call to the logic app backed by the container app.",
```
> a [](http://example.com/codeflow?start=27&length=2)

nit: remove fix typo

razvanbadea-msft commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/LogicApps_Get.json:
```diff
@@ -0,0 +1,19 @@
+        "id": "/subscriptions/8efdecc5-919e-44eb-b179-915dca89ebf9/resourceGroups/examplerg/providers/Microsoft.App/containerApps/testcontainerApp0/providers/Microsoft.App/logicApps/testcontainerApp0",
+        "name": "testcontainerApp0",
+        "type": "Microsoft.App/logicApps",
+        "properties": {}
```
>{} [](http://example.com/codeflow?start=22&length=2)

update this example with more realistic properties

Shubhrmcf-07 replied:
### so far we do not have any properties in the extension body, this has been the placeholder for things we will introduce in the future. Its the same as preview versions, it is already supported in preview with same schema.

gary-x-li replied:
If you don't have any concrete definition, you should remove it. Defining it without any schema and restrict it in the future will break customers. This is especially true for stable versions since you'll need to support it for a long time.

razvanbadea-msft commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,742 @@
+        }
+      ]
+    },
+    "LogicAppProperties": {
```
>LogicAppProperties [](http://example.com/codeflow?start=5&length=18)

How will this be used without a schema

Shubhrmcf-07 replied:
As I mentioned before, this does not have any schema/properties defined as yet, this is supported the same way in preview version and is an object for things we are planning to introduce in properties.

gary-x-li replied:
Similarly as above, this is an anti-pattern.

razvanbadea-msft commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,742 @@
+          "readOnly": true,
+          "description": "Gets the resource type."
+        },
+        "kind": {
```
>kind [](http://example.com/codeflow?start=9&length=4)

what is the difference between kind and type? can it be any string for both?

Shubhrmcf-07 replied:
Type is the resource type, e.g containerApp
Kind specifies a specific kind of containerApp, ex. kind=workflowApp for logic apps

gary-x-li replied:
Please update the description for each property to make it clearer.

Shubhrmcf-07 replied:
updated

razvanbadea-msft commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,742 @@
+        }
+      }
+    },
+    "WorkflowFiles": {
```
>WorkflowFiles [](http://example.com/codeflow?start=5&length=13)

why is this not an array?

Shubhrmcf-07 replied:
This is the same schema as the one we have for web apps, we are migrating to container apps and do not want to change anything. This is not supposed to be an array, it is supposed to be an object with key as filename, and value as file content.

gary-x-li replied:
You need to define a schema for it.

Shubhrmcf-07 replied:
Added detailed description and examples on usage as an object, cannot define schema as the files are sent with filenames as keys and file content as values (which is consistent with our existing offerings on Microsoft.Web)

razvanbadea-msft commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,742 @@
+          "$ref": "#/definitions/WorkflowFiles",
+          "description": "Files of the app."
+        },
+        "filesToDelete": {
```
>filesToDelete [](http://example.com/codeflow?start=9&length=13)

are the files listed here from files. You should expand more the description about this collections and how it will be used

Shubhrmcf-07 replied:
no, this is an array of file names, corresponding file names are checked in the file system and deletes the corresponding files. Again, we are already supporting this extension in preview versions with same schema, and on our existing web offering also this is the same way. This is not a new schema we are introducing.

TimLovellSmith replied:
@Shubhrmcf-07 Did you respond to the concern here?

Shubhrmcf-07 committed:
fixed typo

Shubhrmcf-07 committed:
Added support for kind property in stable versions

Shubhrmcf-07 committed:
Added example ref for container app kind update

Shubhrmcf-07 committed:
Fixed model validation

Shubhrmcf-07 committed:
Included functionapp in kind enums

Shubhrmcf-07 committed:
Added functionapp as example for container app kind

Shubhrmcf-07 committed:
Removed unnecessary space

Shubhrmcf-07 committed:
Updated description for better clarity in workflow envelope

Shubhrmcf-07 committed:
Merge branch 'shusachdeva/addlogicappstostableapi' of https://github.com/Shubhrmcf-07/azure-rest-api-specs into shusachdeva/addlogicappstostableapi

Shubhrmcf-07 committed:
Removed unnecessary property object

Shubhrmcf-07 committed:
Removed unnecessary property object

Shubhrmcf-07 committed:
Schema fixes

Shubhrmcf-07 committed:
Description updates

Shubhrmcf-07 committed:
Updated invoke response to be more descriptive

Schema fixes

Added detailed description for files parameter in deploy artifacts

Prettier fix

Updated description

SDK build fix

SDK build fix

Updated description

Removed unnecessary properties

Removed non-required parameter

rkmanda commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,742 @@
+      },
+      "description": "The workflow filter."
+    },
+    "Object": {
```
If you know its of the form { <filename> : <filecontent> } what is preventing you from defining a type with those 2 properties

Shubhrmcf-07 replied:
it is similar to how other API specs are using it

najian commented:
/azp run

Shubhrmcf-07 committed:
Removed changes from preview version already released

Shubhrmcf-07 commented:
The LintDiff error here is unrelated to the files changed in this PR. Can be ignored.

mentat9 commented:
> The generic object is valid in the sense that we don't have the filename/path and filecontent at the spec. We are using it in 2 places :
> 
> 1.) Represents files in payload of deploy artifacts :
> 
> We do not have the filename/path and file content at spec, so it is represented as just a generic object and can be anything. For example :
> 
> { "files" : { "Stateful1/workflow.json" : {}, "Stateful2/input.txt" : "hello", "Artifacts/test.xml" : "..." }
> 
> so it has to be generic.
> 
> 2.) Represents the response of the invoke API :
> 
> The invoke APIs in logic app are used to invoke runtime operations and can have different outputs based on the headers sent. So there is no definite pattern there. Hence, it is required.
> 
> We have already discussed this spec thoroughly and currently offering it in our preview versions, we do not want to make a lot of changes, with some of the changes in this PR we are already modifying the contract upto a certain extent and not desirable.
> 
> In fact, this is the same contract we are using in our offering on Microsoft.Web and do not want to deter from that.

Was there a modeling discussion @rkmanda requested here: https://github.com/Azure/azure-rest-api-specs/pull/34804#issuecomment-2937889135?

I share the concerns of earlier ARM reviewers. Unschematized JSON is a major ARM anti-pattern and RPs are not allowed to publish APIs that use it anymore, aside from a few scenarios that aren't otherwise practical to schematize. You are creating new APIs with this PR, so current guidance applies: it's not enough that it's more convenient to keep a previously published pattern.

Unschematized objects are vectors for bypassing Azure API versioning. This makes them hard for clients to use. They also can't be supported by several important ARM core features. You need a strong justification to use this pattern in a new API. From what I can tell, the two examples you described above are unlikely to qualify:
1. The first one is easily represented as an array of &lt;string&gt;, &lt;string&gt; pairs (fileName, fileContents). This is fully schematized and supported by all ARM features.
2. The second requires your service to parse the response and return a schematized object representing the result, rather than just dumping the unparsed response to the client. Doing this will make your API better.

If you want to pursue using this pattern, please schedule ARM modeling office hours to go through the scenario and describe the importance of keeping this pattern in a new API. If you have already done that, please provide details.

ajaygit158 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,748 @@
+        "summary": "Gets logic app's connections.",
+        "operationId": "LogicApps_ListWorkflowsConnections",
+        "parameters": [
+          {
```
Why isnt this being modelled as a GET ?

Shubhrmcf-07 replied:
Again, parity with existing offering, and were also advised to keep it this way from container apps side.

najian commented:
/azp run

Shubhrmcf-07 committed:
Merge branch 'shusachdeva/addlogicappstostableapi' of https://github.com/Shubhrmcf-07/azure-rest-api-specs into shusachdeva/addlogicapptostable

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,748 @@
+      "type": "object",
+      "description": "A logic app extension resource",
+      "properties": {
+        "id": {
```
This is not just the resource property, it is being used as the exact body for create request as well as the structured response to multiple other calls, we are using these in response for further computations so need to have this structure (it is the same as in preview version as well - which is already Live since 2024-02-02-preview)

Shubhrmcf-07 committed:
Merge branch 'release-app-Microsoft.App-stable/2025-07-01' into shusachdeva/addlogicappstostableapi

Shubhrmcf-07 committed:
Removed deploy artifacts and invoke from Logic App for remodelling

Shubhrmcf-07 committed:
Merge branch 'shusachdeva/addlogicappstostableapi' of https://github.com/Shubhrmcf-07/azure-rest-api-specs into shusachdeva/addlogicapptostable

Shubhrmcf-07 committed:
Remove unused definitions

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,567 @@
+            "pattern": "^[-\\w\\._\\(\\)]+$"
+          },
+          {
+            "name": "resource",
```
> "name": "resource", [](http://example.com/codeflow?start=11&length=20)

Not sure why change it, you can't really "create or update" by PUT with an empty body. Can you?

Shubhrmcf-07 replied:
Didn't fully understand, this still takes in the ProxyResource as defined in definitions/LogicApp in body, which contains the high level properties like name, id, location. I was asked to remove those parameters as redundant from the ProxyResource.

TimLovellSmith replied:
Sorry looks like in my comment I pasted the wrong text(?!). 
What I meant is, not sure why you are making required false, for the body of the resource.

But I think I understand now, your intention is to make the entire body optional, because there are no required properties.

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/ContainerApps.json:
```diff
@@ -699,6 +702,22 @@
           ],
           "description": "The fully qualified resource ID of the resource that manages this resource. Indicates if this resource is managed by another Azure resource. If this is present, complete mode deployment will not delete the resource if it is removed from the template since it is managed by another resource."
         },
```
since these are not related to spec and just values, we can keep it this way as enum values.

Shubhrmcf-07 replied:
both for control plane and data plane we have ordinal validation for this

Shubhrmcf-07 replied:
even for functionapp (which is not our service), it is the same way. this is how it is in preview as well.

Shubhrmcf-07 committed:
Added enum for defined values

Shubhrmcf-07 committed:
Added enum for defined values

Shubhrmcf-07 committed:
Added enum for defined values

TimLovellSmith commented:
>           "enum": [

nit, docs here could be made more complete
How is client cert forwarded? Where are cert details? Is client certificate checked for validity. when it is accepted, required, or ignored?

---
Refers to: specification/app/resource-manager/Microsoft.App/stable/2025-07-01/ContainerApps.json:1074 in d5c0569. [](commit_id = d5c0569282cf09a99e47e11cf2e346cee4251b0c, deletion_comment = False)

TimLovellSmith commented:
>         },

nit, this doc could be clarified: how are secret names interpreted?

---
Refers to: specification/app/resource-manager/Microsoft.App/stable/2025-07-01/ContainerApps.json:1116 in d5c0569. [](commit_id = d5c0569282cf09a99e47e11cf2e346cee4251b0c, deletion_comment = False)

TimLovellSmith commented:
>         },

nit, the meaning/value/format of this string is unclear to me

---
Refers to: specification/app/resource-manager/Microsoft.App/stable/2025-07-01/ContainerApps.json:1108 in d5c0569. [](commit_id = d5c0569282cf09a99e47e11cf2e346cee4251b0c, deletion_comment = False)

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+            "required": true,
+            "type": "string",
+            "pattern": "^[-\\w\\._\\(\\)]+$"
+          },
```
optional nit, there's a less verbose factoring, you could consider having the parameter definitions in just one place, and using $ref to them - that helps ensure you do them consistently

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+          "type": "string",
+          "readOnly": true,
+          "description": "Gets the resource type for ex. Microsoft.App/logicApps/workflows."
+        },
```
Instead of declaring these properties id/name/type, please inherit ProxyResource.

Shubhrmcf-07 replied:
updated, but started seeing lintdiff errors

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+            "modelAsString": false
+          }
+        },
+        "error": {
```
>error [](http://example.com/codeflow?start=9&length=5)

is it readonly?

Shubhrmcf-07 replied:
yes

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+          "ingress": {
+            "external": true,
+            "targetPort": 80,
+            "allowInsecure": true
```
>allowInsecure [](http://example.com/codeflow?start=13&length=13)

consider putting insecure 'false' instead of true in the examples, since 
a) people are likely to copy them, and we'd like to b) push people into the pit of security success

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+          "activeRevisionsMode": "Single",
+          "ingress": {
+            "external": true,
+            "targetPort": 80,
```
>     "targetPort": 80, [](http://example.com/codeflow?start=7&length=22)

consider making your example show a secure HTTPS configuration instead of HTTP. Or at least having two examples, one secure, one insecure.

TimLovellSmith commented:
>         "customDomains": {

I see its not modeled as x-ms-nullable, so it should be omitted in example responses, not returning null

---
Refers to: specification/app/resource-manager/Microsoft.App/stable/2025-07-01/ContainerApps.json:1029 in d5c0569. [](commit_id = d5c0569282cf09a99e47e11cf2e346cee4251b0c, deletion_comment = False)

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+              "corsPolicy": null,
+              "clientCertificateMode": null,
+              "stickySessions": null,
+              "additionalPortMappings": null
```
'null' value shouldn't be returned in your response unless
your property is annotated nullable

What you would do instead is just omit the value from the response.

Please check your example against your actual service behavior!

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+              "allowInsecure": true,
+              "ipSecurityRestrictions": null,
+              "corsPolicy": null,
+              "clientCertificateMode": null,
```
>      "clientCertificateMode": null, [](http://example.com/codeflow?start=8&length=36)

clientCertificateMode being a non nullable enum value, it should not be null, it should be an actual value, or unspecified.
If you have a default behavior, consider returning the value which best explains the default behavior

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+            "dapr": null,
+            "runtime": null,
+            "maxInactiveRevisions": 100,
+            "service": null
```
do you have examples covering all your optional properties that are null in this example? Its good to have examples covering example usage of all your properties.

Shubhrmcf-07 replied:
we don't need it for logic app, it is container app level properties, our aim with this example is to demonstrate proper payload for updating the kind property, other than that these properties are not ours to update.

TimLovellSmith replied:
@shubhrmcf-07 I'm not sure what you are saying here.

If these properties are not valid in the logic app scenario, it seems to me that you might want to leverage pattern like 'discriminated types', in order to make that more explicit in your swagger.

Shubhrmcf-07 replied:
Removed these as per the other comments and validating against the actual response. The are the container app level properties and as an extension resource on top of container apps this is not ours to update. The example shows the relevant properties that refer to the kind property update, other things are not relevant in this example. I removed those and can be found in other container apps examples

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+          },
+          "template": {
+            "revisionSuffix": "",
+            "terminationGracePeriodSeconds": null,
```
> "terminationGracePeriodSeconds": null, [](http://example.com/codeflow?start=11&length=39)

[ARMBlockingComment] 
again, I think null in responses should require x-ms-nullable on the property

Shubhrmcf-07 replied:
not ours to update, already mentioned as optional but now owned by us: https://github.com/Azure/azure-rest-api-specs/blob/release-app-Microsoft.App-stable/2025-07-01/specification/app/resource-manager/Microsoft.App/stable/2025-07-01/CommonDefinitions.json

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+                }
+              }
+            ],
+            "initContainers": null,
```
>initContainers [](http://example.com/codeflow?start=13&length=14)

where is initContainers declared?

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/LogicApps_ListConnections.json:
```diff
@@ -0,0 +1,44 @@
+                  "connection": {
+                    "id": "string"
+                  },
+                  "connectionRuntimeUrl": "string",
```
>string [](http://example.com/codeflow?start=43&length=6)

a more realistic looking example might be helpful?

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+            "Unknown"
+          ],
+          "type": "string",
+          "x-ms-enum": {
```
yes, getting from runtime

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/ContainerApps_Kind_CreateOrUpdate.json:
```diff
@@ -0,0 +1,200 @@
+              "ipSecurityRestrictions": null,
+              "corsPolicy": null,
+              "clientCertificateMode": null,
+              "stickySessions": null,
```
again, the aim of this file is not to specify all properties of container apps, we are not responsible for putting these values as nullable or getting sample values, these property definitions and examples can already be found in other places and are not ours to update.

Shubhrmcf-07 committed:
removed unreturned properties, minor fixes

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+      "type": "object",
+      "properties": {
+        "extendedCode": {
+          "description": "Type of error.",
```
already documented in -> https://github.com/Azure/azure-rest-api-specs/blob/7757286399f92099f616a02463d7d8a9ad7e70cf/specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/LogicApps_ListWorkflows.json

This is only the error for workflow health, not for the entire call. example documented in above file

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+          "type": "string"
+        },
+        "messageTemplate": {
+          "description": "Message template.",
```
this is actually coming from our runtime and the parameters are already set, its not a whole error response for the call but just a part of the get workflow response, example documented here => https://github.com/Azure/azure-rest-api-specs/blob/7757286399f92099f616a02463d7d8a9ad7e70cf/specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/LogicApps_ListWorkflows.json

Shubhrmcf-07 commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/LogicAppsExtension.json:
```diff
@@ -0,0 +1,576 @@
+          "items": {
+            "type": "string"
+          }
+        },
```
see below

Shubhrmcf-07 committed:
fixed duplicate resource

Commit:
Add example for functionapp

Shubhrmcf-07 committed:
Merge pull request #1 from Shubhrmcf-07/add-kind-functionapp

Add example for functionapp

Shubhrmcf-07 committed:
fix prettier

Shubhrmcf-07 committed:
fix prettier

TimLovellSmith commented:
my mistake, target is development branch

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/stable/2025-07-01/examples/LogicApps_ListConnections.json:
```diff
@@ -0,0 +1,44 @@
+    "200": {
+      "headers": {},
+      "body": {
+        "id": "/subscriptions/8efdecc5-919e-44eb-b179-915dca89ebf9/resourceGroups/examplerg/providers/Microsoft.App/containerApps/testapp2/providers/Microsoft.App/logicApps/testapp2/workflowconfigurations/connections",
```
>/8efdecc5-919e-44eb-b179-915dca89ebf9 [](http://example.com/codeflow?start=29&length=37)

note, I recommend using obviously fake subscription ids though

Shubhrmcf-07 committed:
Updated subscription value to generic guid
